core-war/src/main/java/org/silverpeas/web/jobstartpage/servlets/JobStartPagePeasRequestRouter.java

```diff
@@ -752,6 +768,71 @@ public String getDestination(String function, JobStartPagePeasSessionController
+            return a;
+          };
+          final BiFunction<Collection<DisplaySorted>, JSONCodec.JSONArray, JSONCodec.JSONArray> encodeSortedSubSpaces = (l, a) -> {
+            final String parentId = defaultStringIfNotDefined(jobStartPageSC.getSubSpaceId(),jobStartPageSC.getSpaceId());
+            l.stream()
+                .filter(s -> s.getParentId().equals(parentId))
+                .filter(DisplaySorted::isVisible)
+                .forEach(s -> a.addJSONObject(so -> encodeSortedSpace.apply(Pair.of(s, false), so)));
+            return a;
+          };
+          final BiFunction<Stream<DisplaySorted>, JSONCodec.JSONArray, JSONCodec.JSONArray> encodeSortedPath = (l, a) -> {l
+              .filter(DisplaySorted::isVisible)
+              .forEach(s -> a.addJSONObject(so -> encodeSortedSpace.apply(Pair.of(s, false), so)));
+            return a;
+          };
+          final BiFunction<SpaceInst, JSONCodec.JSONObject, JSONCodec.JSONObject> spaceInstEncoder = (s, so) -> so
+              .put(id, valueOf(s.getLocalId()))
+              .put(label, maintenanceSuffix.apply(s.getId(), s.getName(jobStartPageSC.getLanguage())));
+          final BiFunction<Collection<DisplaySorted>, JSONCodec.JSONArray, JSONCodec.JSONArray> encodeSortedInstances = (p, a) -> {
+            final Stream<DisplaySorted> stream = p != null ? p.stream() : empty();
+            stream.filter(DisplaySorted::isVisible).forEach(i -> a.addJSONObject(so -> so
+                .put(id, i.getId())
+                .put("name", i.getTypeName())
+                .put(label, i.getName())));
+            return a;
+          };
+          o.putJSONArray("selectorSpaces", a -> encodeSortedSpaces.apply(jobStartPageSC.getSpaces(), a));
+          if (isDefined(jobStartPageSC.getSpaceId())) {
+            o.putJSONObject("selectorCurrentSpace", so -> spaceInstEncoder.apply(jobStartPageSC.getSpaceInstById(jobStartPageSC.getSpaceId()), so));
+          }
+          if (isDefined(jobStartPageSC.getSubSpaceId())) {
+            final List<DisplaySorted> path = new ArrayList<>();
+            String parentId = jobStartPageSC.getSubSpaceId();
+            while(isDefined(parentId)) {
+              final DisplaySorted parent = jobStartPageSC.getManagedSpace(parentId);
+              parentId = parent.getParentId();
+              if (isDefined(parentId)) {
+                path.add(0, parent);
+              }
+            }
+            o.putJSONArray("path", a -> encodeSortedPath.apply(path.stream(), a));
+          }
+          o.putJSONArray("spaces", a -> encodeSortedSubSpaces.apply(jobStartPageSC.getSubSpaces(), a));
+          o.putJSONArray("applications", a -> encodeSortedInstances.apply(
+              isDefined(jobStartPageSC.getSubSpaceId()) ?
+                  jobStartPageSC.getSubSpaceComponents() :
+                  jobStartPageSC.getSpaceComponents(), a));
+          return o;
+        }));
```

Comment:
Hum... this JSON construction is a complex code (in term of complexity level). Could it be possible to to have a POJO to represent such a JSON object, to build it with a builder, and then, by using `JSONCodec.encode(T)`, to convert it into a JSON `String` representation?

Reply:
OK, I will export this JSON encoding in a dedicated encoder.